### PR TITLE
[python-package] replace np.find_common_type with np.result_type

### LINKED ISSUE #5990: [python-package] lightgbm 4.0.0 raises a DeprecationWarning with latest numpy 1.25.x due to use of `np.find_common_type`
## Description

I just installed a new environment with latest lightgbm (4.0.0) and numpy (1.25.1) and ran into this issue:

```
C:\ProgramData\Anaconda3\envs\fairlearn\lib\site-packages\lightgbm\sklearn.py:1142: in fit
    super().fit(
C:\ProgramData\Anaconda3\envs\fairlearn\lib\site-packages\lightgbm\sklearn.py:842: in fit
    self._Booster = train(
C:\ProgramData\Anaconda3\envs\fairlearn\lib\site-packages\lightgbm\engine.py:245: in train
    booster = Booster(params=params, train_set=train_set)
C:\ProgramData\Anaconda3\envs\fairlearn\lib\site-packages\lightgbm\basic.py:3096: in __init__
    train_set.construct()
C:\ProgramData\Anaconda3\envs\fairlearn\lib\site-packages\lightgbm\basic.py:2210: in construct
    self._lazy_init(data=self.data, label=self.label, reference=None,
C:\ProgramData\Anaconda3\envs\fairlearn\lib\site-packages\lightgbm\basic.py:1801: in _lazy_init
    data, feature_name, categorical_feature, self.pandas_categorical = _data_from_pandas(data=data,
C:\ProgramData\Anaconda3\envs\fairlearn\lib\site-packages\lightgbm\basic.py:701: in _data_from_pandas
    target_dtype = np.find_common_type(df_dtypes, [])
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

array_types = [<class 'numpy.float64'>, <class 'numpy.int8'>, <class 'numpy.float32'>], scalar_types = []

    @set_module('numpy')
    def find_common_type(array_types, scalar_types):
        """
        Determine common type following standard coercion rules.

        .. deprecated:: NumPy 1.25

            This function is deprecated, use `numpy.promote_types` or
            `numpy.result_type` instead.  To achieve semantics for the
            `scalar_types` argument, use `numpy.result_type` and pass the Python
            values `0`, `0.0`, or `0j`.
            This will give the same results in almost all cases.
            More information and rare exception can be found in the
            `NumPy 1.25 release notes <https://numpy.org/devdocs/release/1.25.0-notes.html>`_.

        Parameters
        ----------
        array_types : sequence
            A list of dtypes or dtype convertible objects representing arrays.
        scalar_types : sequence
            A list of dtypes or dtype convertible objects representing scalars.

        Returns
        -------
        datatype : dtype
            The common data type, which is the maximum of `array_types` ignoring
            `scalar_types`, unless the maximum of `scalar_types` is of a
            different kind (`dtype.kind`). If the kind is not understood, then
            None is returned.

        See Also
        --------
        dtype, common_type, can_cast, mintypecode

        Examples
        --------
        >>> np.find_common_type([], [np.int64, np.float32, complex])
        dtype('complex128')
        >>> np.find_common_type([np.int64, np.float32], [])
        dtype('float64')

        The standard casting rules ensure that a scalar cannot up-cast an
        array unless the scalar is of a fundamentally different kind of data
        (i.e. under a different hierarchy in the data type hierarchy) then
        the array:

        >>> np.find_common_type([np.float32], [np.int64, np.float64])
        dtype('float32')

        Complex is of a different type, so it up-casts the float in the
        `array_types` argument:

        >>> np.find_common_type([np.float32], [complex])
        dtype('complex128')

        Type specifier strings are convertible to dtypes and can therefore
        be used instead of dtypes:

        >>> np.find_common_type(['f4', 'f4', 'i4'], ['c8'])
        dtype('complex128')

        """
        # Deprecated 2022-11-07, NumPy 1.25
>       warnings.warn(
                "np.find_common_type is deprecated.  Please use `np.result_type` "
                "or `np.promote_types`.\n"
                "See https://numpy.org/devdocs/release/1.25.0-notes.html and the "
                "docs for more information.  (Deprecated NumPy 1.25)",
                DeprecationWarning, stacklevel=2)
E       DeprecationWarning: np.find_common_type is deprecated.  Please use `np.result_type` or `np.promote_types`.
E       See https://numpy.org/devdocs/release/1.25.0-notes.html and the docs for more information.  (Deprecated NumPy 1.25)
```
Interestingly, it goes away if I downgrade to <4.0.0.

## Reproducible example

This is from a simple test case we have in the fairlearn repo: `test_smoke[True-LGBMClassifier-demographic_parity-ThresholdOptimizer]`. In short:
```
import numpy as np
import pandas as pd
import pytest
from lightgbm import LGBMClassifier
from fairlearn.postprocessing import ThresholdOptimizer

np.random.seed(0)
n = 100
X0 = np.random.normal(size=n)
X1 = np.random.choice([1, 2, 3], size=n)
Y = np.random.choice([0, 1], size=n)
A = np.random.choice([0, 1], size=n)
df = pd.DataFrame({"X0": X0, "X1": X1})
# Set X1 as categorical
df["X1"] = df["X1"].astype("category")

estimator = LGBMClassifier()
estimator.fit(df, Y)

mitigator = ThresholdOptimizer(
    estimator=estimator,
    constraints="demographic_parity",
    prefit=True,
    predict_method="predict",
)
mitigator.fit(df, Y, sensitive_features=A)
mitigator.predict(df, sensitive_features=A)
```

## Environment info

LightGBM version or commit hash: 4.0.0 (does not happen with <4.0.0)

Command(s) you used to install LightGBM

```shell
pip install lightgbm
```

numpy==1.25.1

### PR DESCRIPTION
**Pull request by jmoralez**
Replaces the use of `np.find_common_type` with `np.result_type` to determine the type that should be used when coercing a pandas dataframe to a numpy array.

This function was added in numpy 1.6 ([docs](https://numpy.org/doc/stable/reference/generated/numpy.result_type.html)), which is from [May 2011](https://pypi.org/project/numpy/1.6.0/) and the lowest version we support [is 1.12](https://github.com/microsoft/LightGBM/blob/8967debeb5cb03875eb96a22dbcaa5387b5abaa3/.ci/test-python-oldest.sh#L10) so there shouldn't be any compatibility issues.

Fixes #5990.

### CODE REVIEW

**Review by jameslamb — APPROVED** [review]
Excellent, thank you so much!

I added the `maintenance` label on this PR since it didn't have one yet, but up to you if you want to categorize it as `fix`. I could go either way.
